include/yaclib/coro/detail/await_on_awaiter.hpp

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <yaclib/async/future.hpp>
+#include <yaclib/coro/coro.hpp>
+#include <yaclib/lazy/task.hpp>
+#include <yaclib/util/detail/atomic_counter.hpp>
+#include <yaclib/util/detail/unique_counter.hpp>
+#include <yaclib/util/type_traits.hpp>
+
+namespace yaclib::detail {
+
+template <bool Single>
+using AwaitOnCounterT =
+  std::conditional_t<Single, OneCounter<NopeBase, NopeDeleter>, AtomicCounter<NopeBase, NopeDeleter>>;
+
+template <bool Single>
+class AwaitOnEvent final : public InlineCore, public AwaitOnCounterT<Single> {
+ public:
+  explicit AwaitOnEvent(std::size_t n) noexcept : AwaitOnCounterT<Single>{n} {
+  }
+
+  BaseCore* _job{nullptr};
+
+ private:
+  template <bool SymmetricTransfer>
+  [[nodiscard]] YACLIB_INLINE auto Impl(InlineCore& caller) noexcept {
+    if constexpr (Single) {
+      _job->_executor->Submit(*_job);
+    } else {
+      if (this->SubEqual(1)) {
+        YACLIB_ASSERT(_job != nullptr);
+        _job->_executor->Submit(*_job);
+      }
+    }
+    return Noop<SymmetricTransfer>();
+  }
+  [[nodiscard]] InlineCore* Here(InlineCore& caller) noexcept final {
+    return Impl<false>(caller);
+  }
+#if YACLIB_SYMMETRIC_TRANSFER != 0
+  [[nodiscard]] yaclib_std::coroutine_handle<> Next(InlineCore& caller) noexcept final {
+    return Impl<true>(caller);
+  }
+#endif
+};
+
+template <bool Single>
+struct [[nodiscard]] AwaitOnAwaiter final {
+  explicit AwaitOnAwaiter(IExecutor& e, BaseCore& caller) noexcept : _executor{e}, _caller{caller}, _event{1} {
+  }
+
+  constexpr bool await_ready() const noexcept {
+    return false;
+  }
+
+  template <typename Promise>
+  YACLIB_INLINE void await_suspend(yaclib_std::coroutine_handle<Promise> handle) noexcept {
+    BaseCore& core = handle.promise();
+    core._executor = &_executor;
+    _event._job = &core;
+
+    if (!_caller.SetCallback(_event)) {
+      _executor.Submit(core);
+    }
+  }
+
+  constexpr void await_resume() const noexcept {
+  }
+
+ private:
+  IExecutor& _executor;
+  BaseCore& _caller;
```

Comment:
You can reuse _job field.
Before await_suspend it's always nullptr and unused.
But in await_suspend you don't really need it.